---
title: "Mojave_Div_ITS_phyloseq"
author: "Nat Pombubpa"
date: "Updated on December 11, 2019"
output: html_document
---

#Fungal taxonomic barplot by site

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###STEP1: Load all necessary packages for analysis
More information about Phyloseq can be found at the following link: [Phyloseq](https://joey711.github.io/phyloseq/)
If you get error in this step, you probably need to install any packages which causes error.

```{r warning=FALSE, message=FALSE}
library(ape)
library(vegan)
library(dplyr)
library(scales)
library(grid)
library(reshape2)
library(phyloseq)
library(magrittr)
library(ggplot2)
library(ggpubr)
library(plyr)
library(data.table)
library(tidyr)
library(tidyverse)
library(multcompView)
library(ggtree)
library(treeio)
library(devtools)
library(MicEco)
library(MoMAColors)
library(castor)
library(tibble)



```



####Initiallize functions
This function was written by Joey (but it is not part of phyloseq
package) <https://github.com/joey711/phyloseq/issues/418>

Kian notes: This defines functions which will be used in future steps. fast_melt takes an input which is a physeq object, extracts OTU table, modifies this (ie removes NA values). summarize_taxa summarizes the relative abundance data by taxonomic rank, but can also be used with grouping variables (ie crust type) and can calculate summary stats. The result is a data.table containing mean, standard deviation, minimum, and maximum relative abundances for each taxonomic group.

```{r warning=FALSE}
fast_melt = function(physeq){
  # supports "naked" otu_table as `physeq` input.
  otutab = as(otu_table(physeq), "matrix")
  if(!taxa_are_rows(physeq)){otutab <- t(otutab)}
  otudt = data.table(otutab, keep.rownames = TRUE)
  setnames(otudt, "rn", "taxaID")
  # Enforce character taxaID key
  otudt[, taxaIDchar := as.character(taxaID)]
  otudt[, taxaID := NULL]
  setnames(otudt, "taxaIDchar", "taxaID")
  # Melt count table
  mdt = melt.data.table(otudt, 
                        id.vars = "taxaID",
                        variable.name = "SampleID",
                        value.name = "count")
  # Remove zeroes, NAs
  mdt <- mdt[count > 0][!is.na(count)]
  # Calculate relative abundance
  mdt[, RelativeAbundance := count / sum(count), by = SampleID]
  if(!is.null(tax_table(physeq, errorIfNULL = FALSE))){
    # If there is a tax_table, join with it. Otherwise, skip this join.
    taxdt = data.table(as(tax_table(physeq, errorIfNULL = TRUE), "matrix"), keep.rownames = TRUE)
    setnames(taxdt, "rn", "taxaID")
    # Enforce character taxaID key
    taxdt[, taxaIDchar := as.character(taxaID)]
    taxdt[, taxaID := NULL]
    setnames(taxdt, "taxaIDchar", "taxaID")
    # Join with tax table
    setkey(taxdt, "taxaID")
    setkey(mdt, "taxaID")
    mdt <- taxdt[mdt]
  }
  return(mdt)
}

summarize_taxa = function(physeq, Rank, GroupBy = NULL){
  Rank <- Rank[1]
  if(!Rank %in% rank_names(physeq)){
    message("The argument to `Rank` was:\n", Rank,
            "\nBut it was not found among taxonomic ranks:\n",
            paste0(rank_names(physeq), collapse = ", "), "\n",
            "Please check the list shown above and try again.")
  }
  if(!is.null(GroupBy)){
    GroupBy <- GroupBy[1]
    if(!GroupBy %in% sample_variables(physeq)){
      message("The argument to `GroupBy` was:\n", GroupBy,
              "\nBut it was not found among sample variables:\n",
              paste0(sample_variables(physeq), collapse = ", "), "\n",
              "Please check the list shown above and try again.")
    }
  }
  # Start with fast melt
  mdt = fast_melt(physeq)
  if(!is.null(GroupBy)){
    # Add the variable indicated in `GroupBy`, if provided.
    sdt = data.table(SampleID = sample_names(physeq),
                     var1 = get_variable(physeq, GroupBy))
    setnames(sdt, "var1", GroupBy)
    # Join
    setkey(sdt, SampleID)
    setkey(mdt, SampleID)
    mdt <- sdt[mdt]
  }
  # Summarize
  Nsamples = nsamples(physeq)
  summarydt = mdt[, list(meanRA = sum(RelativeAbundance)/Nsamples,
                         sdRA = sd(RelativeAbundance),
                         minRA = min(RelativeAbundance),
                         maxRA = max(RelativeAbundance)),
                  by = c(Rank, GroupBy)]
  return(summarydt)
}

plot_taxa_summary = function(physeq, Rank, GroupBy = NULL){
  # Get taxa summary table 
  dt1 = summarize_taxa(physeq, Rank = Rank, GroupBy = GroupBy)
  # Set factor appropriately for plotting
  RankCol = which(colnames(dt1) == Rank)
  setorder(dt1, -meanRA)
  dt1[, RankFac := factor(dt1[[Rank]], 
                          levels = rev(dt1[[Rank]]))]
  dt1[, ebarMax := max(c(0, min(meanRA + sdRA))), by = eval(Rank)]
  dt1[, ebarMin := max(c(0, min(meanRA - sdRA))), by = eval(Rank)]
  # Set zeroes to one-tenth the smallest value
  ebarMinFloor = dt1[(ebarMin > 0), min(ebarMin)]
  ebarMinFloor <- ebarMinFloor / 10
  dt1[(ebarMin == 0), ebarMin := ebarMinFloor]
  
  pRank = ggplot(dt1, aes(x = meanRA, y = RankFac)) +
    scale_x_log10() +
    xlab("Mean Relative Abundance") +
    ylab(Rank) +
    theme_bw()
  if(!is.null(GroupBy)){
    # pRank <- pRank + facet_wrap(facets = as.formula(paste("~", GroupBy)))
    pRank <- pRank + geom_point(mapping = aes_string(colour = GroupBy),
                                size = 5)
  } else {
    # Don't include error bars for faceted version
    pRank <- pRank + geom_errorbarh(aes(xmax = ebarMax,
                                        xmin = ebarMin))
  }
  return(pRank)
}
```

###STEP2: Import Mapping file (metadate file)
1.Check mapping file before import to R, R doesn’t seem to like sample name to start with number or contain “-” in sample name. If you get error in this step, you should check file name first.

2.First column of first row should not start with #, R will not read the first row that starts with #

3. You can choose which samples to include in analysis by indicating "KEEP" in Description column

```{r}
meta = read.table("Metadata_KHK2.tsv",header=TRUE,row.names=1,sep="\t",stringsAsFactors=FALSE)
```

If need to keep or remove some samples
```{r}
#meta <- meta[which(meta$Layer %in% c("Surface")),]
```

###STEP3: Check if your metadata file has been import successfully and correctly

The output will show a table of your metadata file (mapping file).

*If you do not have header, you might start your first row with #

```{r warning=FALSE}
head(meta)
```

###STEP4: Construct sample_data-class using imported metadata

```{r}
sampleData <- sample_data(meta)
```

###STEP5: Import OTU table

OTU table from Mojave 16S data is “MC2017FC.otu_table.txt”.
MC-Mojave Crust-2017-FC-Fungi completed

```{r}
otus <- read.table("Mojave2020_KKelly202307.otu_table.txt",header=T,sep="\t",row.names=1)
otumat <- as(as.matrix(otus), "matrix")
OTU = otu_table(otumat, taxa_are_rows = TRUE)
```

Check imported OTU table

```{r warning=FALSE}
head(OTU)
```

###STEP6: Import taxonomy table
Taxonmy table generated from AMPtk need to be rearranged using following script.

“perl rdp_taxonmy2mat.pl<Input_taxonmy.txt>Output_taxonomy.txt”

rdp_taxonomy2mat.pl was created by Professor Jason E. Stajich

```{r}
taxmat <- read.table("Mojave2020_KKelly202307.ASVs.taxonomy.fix.txt", header=T,sep="\t",row.names=1)
taxmat <- as(as.matrix(taxmat),"matrix")
TAX = tax_table(taxmat)
```

###STEP7: Import phylogenetic tree
Phylogenetic tree can also be include for further phylogenetic analysis.

```{r warning=FALSE}
treefile = "Mojave2020_KKelly202307.ASVs.tree.phy"
tree = read.tree(treefile)
```

###STEP8: Construct Phyloseq object
To construct phyloseq object, otu table, taxonomy table, and sampleData are required. Phylogenetic tree can be included, but it is not necessary for constructing phyloseq object.
Construct Phyloseq object called "Physeq"

```{r warning=FALSE}
physeq = phyloseq(OTU,TAX,sampleData,tree)
```

Check phyloseq object
This should indicate that your physeq is a "phyloseq-class experiment-level object""

```{r warning=FALSE}
physeq
```

###STEP9: Remove singletons
Remove any OTUs that present only one time.

```{r }
physeq.prune = prune_taxa(taxa_sums(physeq) > 1, physeq)
```

```{r warning=FALSE}
physeq.prune
```

###STEP10: Plot read counts to check dataset
Check read counts: any samples that have very low reads should be removed.
[Ref](http://evomics.org/wp-content/uploads/2016/01/phyloseq-Lab-01-Answers.html)

```{r}
readcount = data.table(as(sample_data(physeq.prune), "data.frame"),
                 TotalReads = sample_sums(physeq.prune), 
                 keep.rownames = TRUE)
setnames(readcount, "rn", "SampleID")

#For plotting, use command below.
#SeqDepth = ggplot(readcount, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
```

TotalReads of all the samples can be in this table (select only SampleID and TotalReads columns).
In order to check samples with low number of reads, "order()" can be used to sort "TotalReads" column.
In this dataset, N55.Rhizo has very low number of reads, so will will filter this sample out using the next minimum number of reads.
```{r}
readcount = readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")]
```

```{r}
head(readcount)
```

```{r warning=FALSE}
set.seed(1)
physeq.prune.rarefy = rarefy_even_depth(physeq.prune, sample.size = 7909, replace = FALSE, trimOTUs = FALSE)
physeq.prune.rarefy
```

###STEP11: Getting taxonomy into the tree and labeling

#### Subset taxa for Glomus and Rhizoglomus or Glomeromycetes
```{r}
# Subset taxa for Glomus and Rhizoglomus or Glomeromycetes
#physeq.prune.rarefy.Glomeromycetes <- subset_taxa(physeq.prune.rarefy, Genus %in% c("Glomus", "Rhizophagus"))
physeq.prune.rarefy.Glomeromycetes <- subset_taxa(physeq.prune.rarefy, Class == "Glomeromycetes")


```




#### Extract genus names from the taxonomy table, replace NA

```{r}

tax_table_prune <- tax_table(physeq.prune.rarefy.Glomeromycetes)

species_names <- tax_table_prune[, "Species"]

species_names[is.na(species_names)] <-  tax_table_prune[, "Genus"]

species_names[is.na(species_names)] <-  tax_table_prune[, "Order"]

species_names[is.na(species_names)] <-  tax_table_prune[, "Family"]

species_names[is.na(species_names)] <-  tax_table_prune[, "Class"]


species_names


```

#### replace tip labels with taxonomy

```{r}

# Get the tip labels (ASV IDs) from the phylogenetic tree
tip_labels <- phy_tree(physeq.prune.rarefy.Glomeromycetes)$tip.label

# Match ASV IDs with species names and replace tip labels
new_tip_labels <- species_names[match(tip_labels, rownames(species_names))]

new_tip_labels
```

#### replace tip labels with taxonomy and ASV number

```{r}

fulltree <- phy_tree(physeq.prune.rarefy.Glomeromycetes)

tiplabels_fulltree <- fulltree$tip.label

new_tip_labels_fulltree <- paste(tiplabels_fulltree, new_tip_labels, sep = " ")

fulltree$tip.label <- new_tip_labels_fulltree

fulltree
```


###STEP12: Site counts


####subset by site
```{r}
physeq.prune.rarefy.GlomeromycetesAB <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Site == "AB")

physeq.prune.rarefy.GlomeromycetesODLO <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Site == "ODLO")

physeq.prune.rarefy.GlomeromycetesTP <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Site == "TP")

physeq.prune.rarefy.GlomeromycetesCIMA <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Site == "CIMA")

physeq.prune.rarefy.GlomeromycetesGMT <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Site == "GMT")

```

####sum all taxa


```{r}

#sample_data <- as.data.frame(sample_data(physeq.prune.rarefy.Glomeromycetes))

AB = taxa_sums(physeq.prune.rarefy.GlomeromycetesAB)


TP = taxa_sums(physeq.prune.rarefy.GlomeromycetesTP)


GMT = taxa_sums(physeq.prune.rarefy.GlomeromycetesGMT)


CIMA = taxa_sums(physeq.prune.rarefy.GlomeromycetesCIMA)


ODLO = taxa_sums(physeq.prune.rarefy.GlomeromycetesODLO)

```
####make a df for each site taxa count table with tip labels

```{r}
#Anza Borrego
AB.df <- data.frame(ID=new_tip_labels_fulltree, AB)

AB.df <- AB.df %>%
  mutate(Site = "AB")

colnames(AB.df)[colnames(AB.df) == "AB"] <- "Count"

AB.df
```

```{r}

#Granite mountains

GMT.df <- data.frame(ID=new_tip_labels_fulltree, GMT)

GMT.df <- GMT.df %>%
  mutate(Site = "GMT")

colnames(GMT.df)[colnames(GMT.df) == "GMT"] <- "Count"

GMT.df
```

```{r}

#torrey pines

TP.df <- data.frame(ID=new_tip_labels_fulltree, TP)

TP.df <- TP.df %>%
  mutate(Site = "TP")

colnames(TP.df)[colnames(TP.df) == "TP"] <- "Count"

TP.df
```

```{r}

#Oasis De Los Osos

ODLO.df <- data.frame(ID=new_tip_labels_fulltree, ODLO)

ODLO.df <- ODLO.df %>%
  mutate(Site = "ODLO")

colnames(ODLO.df)[colnames(ODLO.df) == "ODLO"] <- "Count"

ODLO.df
```

```{r}

#CIMA volcanic field

CIMA.df <- data.frame(ID=new_tip_labels_fulltree, CIMA)

CIMA.df <- CIMA.df %>%
  mutate(Site = "CIMA")

colnames(CIMA.df)[colnames(CIMA.df) == "CIMA"] <- "Count"

CIMA.df
```



###STEP13: Get counts by crust_type

####subset by crust_type

```{r}
physeq.prune.rarefy.GlomeromycetesSMC <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Crust_type == "SMC")

physeq.prune.rarefy.GlomeromycetesSMC.ster <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Crust_type == "SMC.ster")

physeq.prune.rarefy.GlomeromycetesDirt <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Crust_type == "Dirt")

physeq.prune.rarefy.GlomeromycetesSMC.sub <- subset_samples(physeq.prune.rarefy.Glomeromycetes, Crust_type == "SMC.sub")

```



####get counts
```{r}
#sample_data <- as.data.frame(sample_data(physeq.prune.rarefy.Glomeromycetes))

SMC = taxa_sums(physeq.prune.rarefy.GlomeromycetesSMC)

Dirt = taxa_sums(physeq.prune.rarefy.GlomeromycetesDirt)

SMC.ster = taxa_sums(physeq.prune.rarefy.GlomeromycetesSMC.ster)

SMC.sub = taxa_sums(physeq.prune.rarefy.GlomeromycetesSMC.sub)

```

####Make dataframes for each Crust_type


```{r}

SMC.df <- data.frame(ID=new_tip_labels_fulltree, SMC)

SMC.df <- SMC.df %>%
  mutate(Crust_type = "SMC")

colnames(SMC.df)[colnames(SMC.df) == "SMC"] <- "Val"

SMC.df
```


```{r}

SMC.ster.df <- data.frame(ID=new_tip_labels_fulltree, SMC.ster)

SMC.ster.df <- SMC.ster.df %>%
  mutate(Crust_type = "SMC.ster")

colnames(SMC.ster.df)[colnames(SMC.ster.df) == "SMC.ster"] <- "Val"

SMC.ster.df
```

```{r}

SMC.sub.df <- data.frame(ID=new_tip_labels_fulltree, SMC.sub)

SMC.sub.df <- SMC.sub.df %>%
  mutate(Crust_type = "SMC.sub")

colnames(SMC.sub.df)[colnames(SMC.sub.df) == "SMC.sub"] <- "Val"
SMC.sub.df
```

```{r}

Dirt.df <- data.frame(ID=new_tip_labels_fulltree, Dirt)

Dirt.df <- Dirt.df %>%
  mutate(Crust_type = "Dirt")

colnames(Dirt.df)[colnames(Dirt.df) == "Dirt"] <- "Val"

Dirt.df
```

###STEP14: Plot tree
#### Generate a palette of colors

```{r}
Dali <- moma.colors("Dali")
OKeeffe <- moma.colors("OKeeffe", n = 9)
Warhol <- moma.colors("Warhol")



# Generate a larger palette of colors
#dali_palette <- moma.colors("Dali", n = 200)

# Subset the palette to include 100 colors
#palette_100 <- dali_palette[1:100]

# Display the first few colors in the palette
#head(palette_100)
```

####Read in Reference phylogeny (See NCBI_ref_phylo pipeline) and remove decimals in ASV

```{r warning=FALSE}
treefile = "Mojave2020.Glomero.fasaln.clipkit.treefile"
tree = read.tree(treefile)
#tree$tip.label = sub("\\..*", "", tree$tip.label)
tree$tip.label <- ifelse(grepl("ASV", tree$tip.label), sub("\\..*", "", tree$tip.label), tree$tip.label)

```

#### build tree and df for tip labels

```{r}


# Root the tree
rooted_tree <- root(tree, outgroup = "Paraglomus_occultum")

to_drop <- c("Mortierella_globalpina", "Archaeospora_ecuadoriana","Acaulospora_entreriana","Acaulospora_laevis","Acaulospora_spinosa","Acaulospora_cavernata","Acaulospora_colombiana", "Paraglomus_occultum")
# Drop the specified tips
straintree <- drop.tip(rooted_tree, to_drop)
# convert tips to df

labels_ref <- as.data.frame(straintree$tip.label)
```


#### build tree and df for tip labels

```{r}

# Get everything that's not an ASV


labels_ref <- as.data.frame(labels_ref[!grepl("ASV", labels_ref$"straintree$tip.label"), ])


# rename column corresponding to tip labels

names(labels_ref)[names(labels_ref) == "labels_ref[!grepl(\"ASV\", labels_ref$\"straintree$tip.label\"), ]"] <- 'labels'

# Make taxonomy and source column for the reference 

labels_ref$Species = "Glomeromycete"

labels_ref$Source = "reference"

# Move labels column so it's rows

row.names(labels_ref) = labels_ref$labels

labels_ref$labels <- NULL


labels_ref
```

#### replace tip labels with taxonomy



```{r}

# Get the tip labels (ASV IDs) from the phylogenetic tree
tip_labels <- phy_tree(physeq.prune.rarefy.Glomeromycetes)$tip.label

# Match ASV IDs with species names and replace tip labels
new_tip_labels <- species_names[match(tip_labels, rownames(species_names))]

new_tip_labels_df <- as.data.frame(new_tip_labels)
new_tip_labels_df
```
```{r}

new_tip_labels_df$Source = "ASV"

combined_tip_labels_Genus <- rbind(new_tip_labels_df, labels_ref)
combined_tip_labels_Genus <- tibble::rownames_to_column(combined_tip_labels_Genus, "ID")
combined_tip_labels_Genus
#right_join(labels_ref, b, by = "x1")
```


#### manually edit table to fix taxonomy in representatives. All MaarjAM only down to family level but NCBI refs down to species is fine


#### replace tip labels with taxonomy and ASV number
```{r}
Levine <- moma.colors("Levine2", n = 9)

p10 <- ggtree(straintree, layout="rectangular") %<+% combined_tip_labels_Genus + geom_tippoint(aes(color=Species, shape=Source), size=1) +
  scale_color_manual(values = Dali) 


plot(p10)
```
###STEP15: Plot faceted tree


```{r}
SMC.df$ID <- NULL


tax_counts_crust_type.SMC <- tibble::rownames_to_column(SMC.df, "ID")

rownames(tax_counts_crust_type.SMC) <- tax_counts_crust_type.SMC[,1]


SMC.ster.df$ID <- NULL

tax_counts_crust_type.SMC.ster <- tibble::rownames_to_column(SMC.ster.df, "ID")

rownames(tax_counts_crust_type.SMC.ster) <- tax_counts_crust_type.SMC.ster[,1]


SMC.sub.df$ID <- NULL


tax_counts_crust_type.SMC.sub <- tibble::rownames_to_column(SMC.sub.df, "ID")

rownames(tax_counts_crust_type.SMC.sub) <- tax_counts_crust_type.SMC.sub[,1]

Dirt.df$ID <- NULL


tax_counts_crust_type.Dirt <- tibble::rownames_to_column(Dirt.df, "ID")

rownames(tax_counts_crust_type.Dirt) <- tax_counts_crust_type.Dirt[,1]


b <- rbind(tax_counts_crust_type.SMC, tax_counts_crust_type.SMC.ster, tax_counts_crust_type.Dirt, tax_counts_crust_type.SMC.sub)


epsilon <- 1  # You can adjust this value based on your preference

b$Val <- log10(b$Val + epsilon)

b
```

```{r}

#labels_ref_tax_counts_crust_typed <- rbind(labels_ref_crust_type, tax_counts_crust_type)
#rownames(labels_ref_tax_counts_crust_type) <- straintree$tip.label
#labels_ref_tax_counts_crust_typed

#labels_ref_tax_counts_crust_type
```

```{r}

p11 <- facet_plot(p10, panel = 'Substrate', data = b, 
				geom = geom_barh, 
				mapping = aes(x = Val, fill = Crust_type), 
				stat='identity' ) +
  scale_fill_manual(values = Dali)
p11

```

```{r}
GMT.df$ID <- NULL


tax_counts_crust_type.GMT.df <- tibble::rownames_to_column(GMT.df, "ID")

rownames(tax_counts_crust_type.GMT.df) <- tax_counts_crust_type.GMT.df[,1]


CIMA.df$ID <- NULL

tax_counts_crust_type.CIMA.df <- tibble::rownames_to_column(CIMA.df, "ID")

rownames(tax_counts_crust_type.CIMA.df) <- tax_counts_crust_type.CIMA.df[,1]


ODLO.df$ID <- NULL


tax_counts_crust_type.ODLO.df <- tibble::rownames_to_column(ODLO.df, "ID")

rownames(tax_counts_crust_type.ODLO.df) <- tax_counts_crust_type.ODLO.df[,1]

AB.df$ID <- NULL


tax_counts_crust_type.AB.df <- tibble::rownames_to_column(AB.df, "ID")

rownames(tax_counts_crust_type.AB.df) <- tax_counts_crust_type.AB.df[,1]

TP.df$ID <- NULL


tax_counts_crust_type.TP.df <- tibble::rownames_to_column(TP.df, "ID")

rownames(tax_counts_crust_type.TP.df) <- tax_counts_crust_type.TP.df[,1]


d <- rbind(tax_counts_crust_type.AB.df, tax_counts_crust_type.TP.df, tax_counts_crust_type.GMT.df, tax_counts_crust_type.ODLO.df, tax_counts_crust_type.CIMA.df)


epsilon <- 1  # You can adjust this value based on your preference

d$Count <- log10(d$Count + epsilon)

d
```

```{r}
OKeeffe <- moma.colors("VanGogh", n = 9)


p12 <- facet_plot(p11, panel = 'Site', data = d, 
				geom = geom_barh, 
				mapping = aes(x = Count, fill = Site), 
				stat='identity' ) + guides(fill=guide_legend(title="Category")) + 
  scale_fill_manual(values = Dali, breaks = c("AB", "CIMA", "GMT", "ODLO", "TP", "Dirt", "SMC","SMC.ster","SMC.sub"), labels = c("AB", "CIMA", "GMT", "ODLO", "TP", "Dirt", "SMC","SMC.ster","SMC.sub"))
p12

```

```{r}

# Save the heatmap as a file
ggsave("glomeromycetebarplot_NCBI_ref.png", plot = p12, width = 17, height = 9)

```

```{r}

p11 <- facet_plot(p10, panel = 'Site', data = d, 
				geom = geom_barh, 
				mapping = aes(x = Count, fill = Site), 
				stat='identity' ) +
  scale_fill_manual(values = Dali)
p11

```