---
title: "Mojave_Div_ITS_phyloseq"
author: "Nat Pombubpa"
date: "Updated on December 11, 2019"
output: html_document
---

#Fungal taxonomic barplot by site

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###STEP1: Load all necessary packages for analysis
More information about Phyloseq can be found at the following link: [Phyloseq](https://joey711.github.io/phyloseq/)
If you get error in this step, you probably need to install any packages which causes error.

```{r warning=FALSE, message=FALSE}
library(ape)
library(vegan)
library(dplyr)
library(scales)
library(grid)
library(reshape2)
library(phyloseq)
library(magrittr)
library(ggplot2)
library(ggpubr)
library(plyr)
library(data.table)
library(tidyr)
library(tidyverse)
library(multcompView)
library(ggtree)
library(treeio)
library(devtools)
library(MicEco)
```

###STEP2: Import Mapping file (metadate file)
1.Check mapping file before import to R, R doesn’t seem to like sample name to start with number or contain “-” in sample name. If you get error in this step, you should check file name first.

2.First column of first row should not start with #, R will not read the first row that starts with #

3. You can choose which samples to include in analysis by indicating "KEEP" in Description column

```{r}
meta = read.table("metadata_khk.tsv",header=TRUE,row.names=1,sep="\t",stringsAsFactors=FALSE)
```

If need to keep or remove some samples
```{r}
#meta <- meta[which(meta$Layer %in% c("Surface")),]
```

###STEP3: Check if your metadata file has been import successfully and correctly

The output will show a table of your metadata file (mapping file).

*If you do not have header, you might start your first row with #

```{r warning=FALSE}
head(meta)
```

###STEP4: Construct sample_data-class using imported metadata

```{r}
sampleData <- sample_data(meta)
```

###STEP5: Import OTU table

OTU table from Mojave 16S data is “MC2017FC.otu_table.txt”.
MC-Mojave Crust-2017-FC-Fungi completed

```{r}
otus <- read.table("Mojave2020_KKelly202307.otu_table.txt",header=T,sep="\t",row.names=1)
otumat <- as(as.matrix(otus), "matrix")
OTU = otu_table(otumat, taxa_are_rows = TRUE)
```

Check imported OTU table

```{r warning=FALSE}
head(OTU)
```

###STEP6: Import taxonomy table
Taxonmy table generated from AMPtk need to be rearranged using following script.

“perl rdp_taxonmy2mat.pl<Input_taxonmy.txt>Output_taxonomy.txt”

rdp_taxonomy2mat.pl was created by Professor Jason E. Stajich

```{r}
taxmat <- read.table("Mojave2020_KKelly202307.ASVs.taxonomy.fix.txt", header=T,sep="\t",row.names=1)
taxmat <- as(as.matrix(taxmat),"matrix")
TAX = tax_table(taxmat)
```

###STEP7: Import phylogenetic tree
Phylogenetic tree can also be include for further phylogenetic analysis.

```{r warning=FALSE}
treefile = "Mojave2020_KKelly202307.ASVs.tree.phy"
tree = read.tree(treefile)
```

###STEP8: Construct Phyloseq object
To construct phyloseq object, otu table, taxonomy table, and sampleData are required. Phylogenetic tree can be included, but it is not necessary for constructing phyloseq object.
Construct Phyloseq object called "Physeq"

```{r warning=FALSE}
physeq = phyloseq(OTU,TAX,sampleData,tree)
```

Check phyloseq object
This should indicate that your physeq is a "phyloseq-class experiment-level object""

```{r warning=FALSE}
physeq
```

###STEP9: Remove singletons
Remove any OTUs that present only one time.

```{r }
physeq.prune = prune_taxa(taxa_sums(physeq) > 1, physeq)
```

```{r warning=FALSE}
physeq.prune
```

###STEP10: Plot read counts to check dataset
Check read counts: any samples that have very low reads should be removed.
[Ref](http://evomics.org/wp-content/uploads/2016/01/phyloseq-Lab-01-Answers.html)

```{r}
readcount = data.table(as(sample_data(physeq.prune), "data.frame"),
                 TotalReads = sample_sums(physeq.prune), 
                 keep.rownames = TRUE)
setnames(readcount, "rn", "SampleID")

#For plotting, use command below.
#SeqDepth = ggplot(readcount, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
```

TotalReads of all the samples can be in this table (select only SampleID and TotalReads columns).
In order to check samples with low number of reads, "order()" can be used to sort "TotalReads" column.
In this dataset, N55.Rhizo has very low number of reads, so will will filter this sample out using the next minimum number of reads.
```{r}
readcount = readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")]
```

```{r}
head(readcount)
```

```{r warning=FALSE}
set.seed(1)
physeq.prune.rarefy = rarefy_even_depth(physeq.prune, sample.size = 7909, replace = FALSE, trimOTUs = FALSE)
physeq.prune.rarefy
```

##Tree for Glom

```{r}
physeq.prune.rarefy.Glomeromycetes = subset_taxa(physeq.prune.rarefy, Class == "Glomeromycetes")
physeq.prune.rarefy.Glomeromycetes
```
```{r}
physeq.prune.rarefy.Cryptococcus = subset_taxa(physeq.prune.rarefy, Genus == "Rhodotorula")
physeq.prune.rarefy.Cryptococcus
```


```{r}
p1 <- plot_tree(physeq.prune.rarefy.Glomeromycetes, color="Crust_type", shape="Site", label.tips="Genus", size="abundance", plot.margin=0.3)
p1
```

```{r}
p1 <- plot_tree(physeq.prune.rarefy.Glomeromycetes, color="Crust_type", shape="Site", label.tips="Genus", size="abundance", plot.margin=0.3)
p1
```

```{r}

tax_table(physeq.prune.rarefy.Glomeromycetes)

```

```{r}

phy_tree(physeq.prune.rarefy.Glomeromycetes)



# Plot the modified tree
p2 <- ggtree(phy_tree(physeq.prune.rarefy.Glomeromycetes), layout="rectangular") +
  geom_tiplab(size=2, color="black")
p2

tips <- p2$tip.label

```
```{r}
# Define the number of colors you want
nb.cols <- 21
mycolors <- colorRampPalette(brewer.pal(10, "Set3"))(nb.cols)
# Create a ggplot with 21 colors
# Use scale_fill_manual

pal <- wes_palette("Zissou1", 100, type = "continuous")

```


```{r}
# Create heatmap

p_heatmap <- gheatmap(p2, (otu_table(physeq.prune.rarefy.Glomeromycetes)), width=1.5, offset=0.3, 
                      legend_title="Relative Abundance", colnames_position="bottom", 
                      colnames_angle=45, colnames_offset_y=0, hjust=1, font.size=1) +
  scale_fill_gradientn("Relative abundance", colours = pal) +
  scale_y_continuous(limits=c(0, 40)) + theme_tree(legend.position="right")
plot(p_heatmap)

#find 
```
```{r}

# Extract species names from the taxonomy table
species_names <- tax_table(physeq.prune.rarefy.Glomeromycetes)[, "Genus"]

species_names[is.na(species_names)] <- "Unknown AMF"


# Get the tip labels (ASV IDs) from the phylogenetic tree
tip_labels <- phy_tree(physeq.prune.rarefy.Glomeromycetes)$tip.label

# Match ASV IDs with species names and replace tip labels
new_tip_labels <- species_names[match(tip_labels, rownames(species_names))]

new_tip_labels
```
```{r}

object <- phy_tree(physeq.prune.rarefy.Glomeromycetes)

tiplabels <- object$tip.label

new_tip_labels <- paste(tiplabels, new_tip_labels, sep = " ")

object$tip.label <- new_tip_labels

object
```

```{r}

# Plot the modified tree with species names as tip labels
p3 <- ggtree(object, layout="rectangular") +
  geom_tiplab(size=2, color="black")


p3
```
```{r}
physeq.prune.rarefy.Glomeromycetes
```

```{r}
# Get the OTU table
otu_table <- otu_table(physeq.prune.rarefy.Glomeromycetes)

species_names[is.na(species_names)] <- "Unknown Glomeromycete"

# Replace row names in the OTU table with species names
rownames(otu_table) <- new_tip_labels

# Get the sample data from the phyloseq object
sample_data <- as.data.frame(sample_data(physeq.prune.rarefy.Glomeromycetes))

# Extract the variables you want to use as column names (e.g., soil type, site)
column_names <- paste(sample_data$Crust_type, sample_data$Site, sep = " ")



log_otu_table <- log10(otu_table + 1)  # Adding 1 to avoid log(0)

# Replace column names in the OTU table
colnames(log_otu_table) <- column_names

colnames(log_otu_table) <- make.unique(colnames(log_otu_table))


p_heatmap <- gheatmap(p3, log_otu_table, width=1.5, offset=0.3, 
                      legend_title="Relative Abundance", colnames_position="bottom", 
                      colnames_angle=45, colnames_offset_y=0, hjust=1, font.size=1.5) +
  scale_fill_gradientn("Relative abundance", colours = pal) +
  scale_y_continuous(limits=c(-2, 32)) + theme_tree(legend.position="right")
plot(p_heatmap)
```

```{r}

# Save the heatmap as a file
ggsave("glomeromycete.png", plot = p_heatmap, width = 8, height = 6)

```



```{r}

# Extract species names from the taxonomy table
species_names <- tax_table(physeq.prune.rarefy.Cryptococcus)[, "Species"]

species_names[is.na(species_names)] <- "Unknown Rhodotorula sp."

# Get the tip labels (ASV IDs) from the phylogenetic tree
tip_labels <- phy_tree(physeq.prune.rarefy.Cryptococcus)$tip.label

# Match ASV IDs with species names and replace tip labels
new_tip_labels <- species_names[match(tip_labels, rownames(species_names))]

new_tip_labels
```
```{r}

object <- phy_tree(physeq.prune.rarefy.Cryptococcus)

tiplabels <- object$tip.label

new_tip_labels <- paste(tiplabels, new_tip_labels, sep = " ")

object$tip.label <- new_tip_labels

object
```

```{r}

# Plot the modified tree with species names as tip labels
p3 <- ggtree(object, layout="rectangular") +
  geom_tiplab(size=2, color="black")

p3
```
```{r}

# Get the OTU table
otu_table <- otu_table(physeq.prune.rarefy.Cryptococcus)

# Replace row names in the OTU table with species names
rownames(otu_table) <- new_tip_labels

# Take the log10 of the OTU table values
log_otu_table <- log10(otu_table + 1)  # Adding 1 to avoid log(0)

# Replace column names in the OTU table
colnames(log_otu_table) <- column_names

colnames(log_otu_table) <- make.unique(colnames(log_otu_table))

p_heatmap <- gheatmap(p3, log_otu_table, width=1.5, offset=0.2, 
                      legend_title="Relative Abundance", colnames_position="bottom", 
                      colnames_angle=45, colnames_offset_y=0, hjust=1, font.size=1.5) +
  scale_fill_gradientn("Relative abundance", colours = pal) +
  scale_y_continuous(limits=c(-3, 27)) + theme_tree(legend.position="right")
plot(p_heatmap)
```

```{r}
# Create a vector containing both "SMC" and "SMC.ster"
crust_types <- c("SMC.ster", "SMC")

# Subset the samples based on the crust types
physeq.prune.rarefy.crust_type <- subset_samples(physeq.prune.rarefy, Crust_type %in% crust_types)

physeq.prune.rarefy.crust_type
```
```{r}
library("DESeq2")

diagdds = phyloseq_to_deseq2(physeq.prune.rarefy.crust_type, ~ Crust_type)
diagdds = DESeq(diagdds, test="Wald", fitType="parametric")

diagdds
```

```{r}
res = results(diagdds, cooksCutoff = FALSE)
alpha = 0.01
sigtab = res[which(res$padj < alpha), ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(physeq.prune.rarefy.crust_type)[rownames(sigtab), ], "matrix"))
head(sigtab)
```

```{r}

# Save the heatmap as a file
ggsave("Rhodo.png", plot = p_heatmap, width = 8, height = 6)

```

```{r}
library("ggplot2")
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
    scale_fill_brewer(palette = palname, ...)
}

# Filter out rows with unknown genus
sigtab_filtered <- subset(sigtab, !is.na(Genus) & Genus != "")

# Phylum order
x = tapply(sigtab_filtered$log2FoldChange, sigtab_filtered$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab_filtered$Phylum = factor(as.character(sigtab_filtered$Phylum), levels=names(x))

# Genus order
x = tapply(sigtab_filtered$log2FoldChange, sigtab_filtered$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab_filtered$Genus = factor(as.character(sigtab_filtered$Genus), levels=names(x))

log2plot <- ggplot(sigtab_filtered, aes(x=Genus, y=log2FoldChange, color=Class)) + 
  geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5),
        panel.grid.major.x = element_blank(),  # Remove major x-axis gridlines
        panel.grid.minor.x = element_blank(),  # Remove minor x-axis gridlines
        panel.grid.minor.y = element_blank())  # Remove minor x-axis gridlines
log2plot

```
```{r}
# Create a vector containing both "SMC" and "SMC.ster"
Site <- c("ODLO","AB")

# Subset the samples based on the crust types
physeq.prune.rarefy.crust_type <- subset_samples(physeq.prune.rarefy, Site %in% Site)

physeq.prune.rarefy.crust_type
```

```{r}
library("DESeq2")

diagdds = phyloseq_to_deseq2(physeq.prune.rarefy.crust_type, ~ Crust_type)
diagdds = DESeq(diagdds, test="Wald", fitType="parametric")

diagdds
```

```{r}
res = results(diagdds, cooksCutoff = FALSE)
alpha = 0.01
sigtab = res[which(res$padj < alpha), ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(physeq.prune.rarefy.crust_type)[rownames(sigtab), ], "matrix"))
head(sigtab)
```

```{r}
library("ggplot2")
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
    scale_fill_brewer(palette = palname, ...)
}

# Filter out rows with unknown genus
sigtab_filtered <- subset(sigtab, !is.na(Genus) & Genus != "")

# Phylum order
x = tapply(sigtab_filtered$log2FoldChange, sigtab_filtered$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab_filtered$Phylum = factor(as.character(sigtab_filtered$Phylum), levels=names(x))

# Genus order
x = tapply(sigtab_filtered$log2FoldChange, sigtab_filtered$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab_filtered$Genus = factor(as.character(sigtab_filtered$Genus), levels=names(x))

log2plot <- ggplot(sigtab_filtered, aes(x=Genus, y=log2FoldChange, color=Class)) + 
  geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5),
        panel.grid.major.x = element_blank(),  # Remove major x-axis gridlines
        panel.grid.minor.x = element_blank(),  # Remove minor x-axis gridlines
        panel.grid.minor.y = element_blank())  # Remove minor x-axis gridlines
log2plot

```


```{r}

# Save the heatmap as a file
ggsave("Rhodo.png", plot = p_heatmap, width = 8, height = 6)

```

```{r}

# Save the heatmap as a file
ggsave("Rhodo.png", plot = log2plot, width = 8, height = 6)

```

```{r}

ps_venn(physeq.prune.rarefy, "Crust_type", quantities = list(type=c("counts"), font = 1), labels = list(cex = 1.5), col = "black", fill = c("red","blue","green","violet"))

```

```{r}

ps_venn(physeq.prune.rarefy, "Site", quantities = list(type=c("counts"), font = 1), labels = list(cex = 1.5), col = "black", fill = c("red","blue","green","violet"), relative = TRUE)

```
```{r}

ps_pheatmap(
  physeq.prune.rarefy,
  annot_samp = NULL,
  annot_taxa = NULL,
  relative = TRUE,
  log10 = TRUE,
  tax_agg = NULL,
  order_taxa = TRUE,
  min_samples = 1,
  min_reads = 1,
  min_abundance = 0,
  label_rank = NULL,
  color = c("black", viridis::viridis(10)),)
```
##Subset for black yeast


```{r}
physeq.prune.rarefy = subset_taxa(physeq.prune.rarefy, Order=="Chaetothyriales" | Order=="Capnodiales" | Order=="Dothideales" | Order=="Pleosporales" )
physeq.prune.rarefy
```

```{r}
physeq.prune.rarefy.SF = subset_samples(physeq.prune.rarefy, Layer=="Surface")
```

##Make taxonomy table into a matrix and relabel NA as unknown
```{r}
tax.fun <- as(tax_table(physeq.prune.rarefy.SF),"matrix")
head(tax.fun)
tax.fun[is.na(tax.fun)] <- "Unknown"
head(tax.fun)
```

###Convert tax table back to phyloseq object and generate phyloseq object with new tax table
```{r}
TAX.fun <- tax_table(tax.fun)
fun.3 <- phyloseq(sample_data(physeq.prune.rarefy.SF),otu_table(physeq.prune.rarefy.SF),TAX.fun)
fun.3
```

```{r}
glom.fun <- tax_glom(fun.3,taxrank = "Genus")
glom.fun
```

```{r}
tax_table(glom.fun)
```

##Transform OTU table to show relative abundance
##Samples can also be merged together by a variable in the mapping file

```{r}
fun.abund <- merge_samples(glom.fun, "Site")
sample_data(fun.abund)$Site <- factor(sample_names(fun.abund))
fun.abund = transform_sample_counts(fun.abund, function(x) x / sum(x))
fun.abund
```

##Merge taxonomic data with OTU table and mapping file (Can alter code to change taxonomic rank to Order, Class, Family, etc.) and change Phylum column from a factor to a character.

```{r}
data_glom.fun <- psmelt(fun.abund)
data_glom.fun$Genus <- as.character(data_glom.fun$Genus)
```

##If a phylum has less than 1% abundance, phylum name is changed to <1% abund.

```{r}
data_glom.fun$Genus[data_glom.fun$Abundance < 0.01] <- "<1% abund."
```

Count the levels present in the Phylum column

```{r}
Count = length(unique(data_glom.fun$Genus))
Count
```

Print out unique phyla names for insertion into barplots in next step.

```{r}
unique((data_glom.fun$Genus))
```

Create levels of phyla represented in barplot. Levels appear on the barplot in the order they are listed

```{r}
data_glom.fun$Genus <- factor(data_glom.fun$Genus, levels = c("Unknown", "Coniosporium", "Preussia", "Cladophialophora", "Neophaeococcomyces", "Alternaria", "Leptosphaerulina", "Macroventuria", "Sclerostagonospora", "Corynespora", "Chaetosphaeronema", "Paraphoma", "<1% abund."))
```

###Create barplot of phyla - use facet_grid to separate samples by a variable ie in mapping file. Choose colors for plot in scale_fill_manual. color names can be found at http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf *amount of colors must match amount of level in phyla column

```{r}
data_glom.fun$Site = factor(data_glom.fun$Site, levels = c("AB", "CIMA", "ODLO"))
```

```{r fig.height=5, fig.width=8, fig.align="center"}
# The palette with black:
cbbPalette <- c("#009E73","#999999", "#E69F00", "#56B4E9", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "midnightblue", "lightgreen", "aquamarine4","lavenderblush2","blue", "snow3")
taxcom_layer_pl = ggplot(data = data_glom.fun, mapping = aes_string(x = "Site" ,y = "Abundance", fill = "Genus" )) + 
  geom_bar(stat="identity", position="fill") + 
  ggtitle("Black Yeast Taxonomic Composition (Genus level) by Site")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_fill_manual(values = cbbPalette)
print(taxcom_layer_pl)
```

```{r}
png("./Figures/FigS1A_Black_Yeast_Taxonomic_Composition_(Genus_level)_by_Site_not_Rainbow.png", units="in", width = 5.8, height = 5.8, res = 300 )
taxcom_layer_pl
dev.off()
```

```{r}
physeq.prune.rarefy.Glomeromycetes = subset_taxa(physeq.prune.rarefy, Genus == "Coniosporium")
physeq.prune.rarefy.Glomeromycetes
```



```{r}
p1 <- plot_tree(physeq.prune.rarefy.Glomeromycetes, color="Crust_type", shape="Site", label.tips="OTU", size="abundance", plot.margin=0.3)
p1
```
```{r}

otu_table(physeq.prune.rarefy.Glomeromycetes)

```

```{r}

phy_tree(physeq.prune.rarefy.Glomeromycetes)

# Plot the modified tree
p2 <- ggtree(phy_tree(physeq.prune.rarefy.Glomeromycetes), layout="rectangular") +
  geom_tiplab(size=2, color="black")
p2
```
```{r}
# Define the number of colors you want
nb.cols <- 21
mycolors <- colorRampPalette(brewer.pal(10, "Set3"))(nb.cols)
# Create a ggplot with 21 colors
# Use scale_fill_manual

pal <- wes_palette("Zissou1", 100, type = "continuous")

```

```{r}
# Create heatmap
p_heatmap <- gheatmap(p2, (otu_table(physeq.prune.rarefy.Glomeromycetes)), width=1.5, offset=0.3, 
                      legend_title="Relative Abundance", colnames_position="bottom", 
                      colnames_angle=45, colnames_offset_y=0, hjust=1, font.size=1) +
  scale_fill_gradientn("Relative abundance", colours = pal) +
  scale_y_continuous(limits=c(0, 60)) + theme_tree(legend.position="right")
plot(p_heatmap)
```






