---
title: "Mojave_Colorado_Div_ITS_phyloseq"
author: "Mostly Nuttapon Pombupba, a bit by Kian Kelly"
date: "Updated on December 21, 2023"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###STEP1: Load all necessary packages

```{r warning=FALSE, message=FALSE}
library(ape)
library(vegan)
library(dplyr)
library(scales)
library(grid)
library(reshape2)
library(phyloseq)
library(magrittr)
library(ggplot2)
library(ggpubr)
library(plyr)
library(data.table)
library(tidyr)
library(tidyverse)
library(multcompView)
library(MoMAColors)

```

####Initiallize functions
This function was written by Joey (but it is not part of phyloseq
package) <https://github.com/joey711/phyloseq/issues/418>

Kian notes: This defines functions which will be used in future steps. fast_melt takes an input which is a physeq object, extracts OTU table, modifies this (ie removes NA values). summarize_taxa summarizes the relative abundance data by taxonomic rank, but can also be used with grouping variables (ie crust type) and can calculate summary stats. The result is a data.table containing mean, standard deviation, minimum, and maximum relative abundances for each taxonomic group.

```{r warning=FALSE}
fast_melt = function(physeq){
  # supports "naked" otu_table as `physeq` input.
  otutab = as(otu_table(physeq), "matrix")
  if(!taxa_are_rows(physeq)){otutab <- t(otutab)}
  otudt = data.table(otutab, keep.rownames = TRUE)
  setnames(otudt, "rn", "taxaID")
  # Enforce character taxaID key
  otudt[, taxaIDchar := as.character(taxaID)]
  otudt[, taxaID := NULL]
  setnames(otudt, "taxaIDchar", "taxaID")
  # Melt count table
  mdt = melt.data.table(otudt, 
                        id.vars = "taxaID",
                        variable.name = "SampleID",
                        value.name = "count")
  # Remove zeroes, NAs
  mdt <- mdt[count > 0][!is.na(count)]
  # Calculate relative abundance
  mdt[, RelativeAbundance := count / sum(count), by = SampleID]
  if(!is.null(tax_table(physeq, errorIfNULL = FALSE))){
    # If there is a tax_table, join with it. Otherwise, skip this join.
    taxdt = data.table(as(tax_table(physeq, errorIfNULL = TRUE), "matrix"), keep.rownames = TRUE)
    setnames(taxdt, "rn", "taxaID")
    # Enforce character taxaID key
    taxdt[, taxaIDchar := as.character(taxaID)]
    taxdt[, taxaID := NULL]
    setnames(taxdt, "taxaIDchar", "taxaID")
    # Join with tax table
    setkey(taxdt, "taxaID")
    setkey(mdt, "taxaID")
    mdt <- taxdt[mdt]
  }
  return(mdt)
}

summarize_taxa = function(physeq, Rank, GroupBy = NULL){
  Rank <- Rank[1]
  if(!Rank %in% rank_names(physeq)){
    message("The argument to `Rank` was:\n", Rank,
            "\nBut it was not found among taxonomic ranks:\n",
            paste0(rank_names(physeq), collapse = ", "), "\n",
            "Please check the list shown above and try again.")
  }
  if(!is.null(GroupBy)){
    GroupBy <- GroupBy[1]
    if(!GroupBy %in% sample_variables(physeq)){
      message("The argument to `GroupBy` was:\n", GroupBy,
              "\nBut it was not found among sample variables:\n",
              paste0(sample_variables(physeq), collapse = ", "), "\n",
              "Please check the list shown above and try again.")
    }
  }
  # Start with fast melt
  mdt = fast_melt(physeq)
  if(!is.null(GroupBy)){
    # Add the variable indicated in `GroupBy`, if provided.
    sdt = data.table(SampleID = sample_names(physeq),
                     var1 = get_variable(physeq, GroupBy))
    setnames(sdt, "var1", GroupBy)
    # Join
    setkey(sdt, SampleID)
    setkey(mdt, SampleID)
    mdt <- sdt[mdt]
  }
  # Summarize
  Nsamples = nsamples(physeq)
  summarydt = mdt[, list(meanRA = sum(RelativeAbundance)/Nsamples,
                         sdRA = sd(RelativeAbundance),
                         minRA = min(RelativeAbundance),
                         maxRA = max(RelativeAbundance)),
                  by = c(Rank, GroupBy)]
  return(summarydt)
}

plot_taxa_summary = function(physeq, Rank, GroupBy = NULL){
  # Get taxa summary table 
  dt1 = summarize_taxa(physeq, Rank = Rank, GroupBy = GroupBy)
  # Set factor appropriately for plotting
  RankCol = which(colnames(dt1) == Rank)
  setorder(dt1, -meanRA)
  dt1[, RankFac := factor(dt1[[Rank]], 
                          levels = rev(dt1[[Rank]]))]
  dt1[, ebarMax := max(c(0, min(meanRA + sdRA))), by = eval(Rank)]
  dt1[, ebarMin := max(c(0, min(meanRA - sdRA))), by = eval(Rank)]
  # Set zeroes to one-tenth the smallest value
  ebarMinFloor = dt1[(ebarMin > 0), min(ebarMin)]
  ebarMinFloor <- ebarMinFloor / 10
  dt1[(ebarMin == 0), ebarMin := ebarMinFloor]
  
  pRank = ggplot(dt1, aes(x = meanRA, y = RankFac)) +
    scale_x_log10() +
    xlab("Mean Relative Abundance") +
    ylab(Rank) +
    theme_bw()
  if(!is.null(GroupBy)){
    # pRank <- pRank + facet_wrap(facets = as.formula(paste("~", GroupBy)))
    pRank <- pRank + geom_point(mapping = aes_string(colour = GroupBy),
                                size = 5)
  } else {
    # Don't include error bars for faceted version
    pRank <- pRank + geom_errorbarh(aes(xmax = ebarMax,
                                        xmin = ebarMin))
  }
  return(pRank)
}
```

###STEP2: Import Mapping file (metadate file) 

Kian notes:

I excluded Nat's samples from this analysis by removing them from the metadata file. 

```{r}
meta = read.table("metadata.tsv",header=TRUE,row.names=1,sep="\t",stringsAsFactors=FALSE)
```

###STEP3: Check if your metadata file has been imported successfully 

```{r warning=FALSE}
head(meta)
```

###STEP4: Construct sample_data-class using imported metadata

Kian notes: basically makes a phyloseq version of your metadata file
```{r}
sampleData <- sample_data(meta)
```

###STEP5: Import OTU table
Kian notes: OTU table here combines my data with Nat's.
NOTE: IF THE HEADER IS IMPROPERLY READ, REMOVE THE COMMA ADDED BY PHYLOSEQ!!!
```{r}
getwd()
otus <- read.table("Mojave2020_KKelly202307.otu_table.txt",header=T,sep="\t",row.names=1)
otumat <- as(as.matrix(otus), "matrix")
OTU = otu_table(otumat, taxa_are_rows = TRUE)
```

Check imported OTU table

```{r warning=FALSE}
head(OTU)
```

###STEP6: Import taxonomy table Taxonmy 
Note: table generated from AMPtk needs to be rearranged using a perl script.

"perl rdp_taxonmy2mat.pl\<Input_taxonmy.txt\>Output_taxonomy.txt"

rdp_taxonomy2mat.pl was created by Professor Jason E. Stajich

```{r}
taxmat <- read.table("Mojave2020_KKelly202307.ASVs.taxonomy.fix.txt", header=T,sep="\t",row.names=1)
taxmat <- as(as.matrix(taxmat),"matrix")
TAX = tax_table(taxmat)
head(TAX)
```

###STEP7: Import phylogenetic tree

```{r warning=FALSE}
#treefile = "Mojave2020_KKelly202307.ASVs.tree.phy"
#tree = read.tree(treefile)
```

###STEP8: Construct Phyloseq object 

To construct phyloseq object, otu
table, taxonomy table, and sampleData are required. Phylogenetic tree
can be included, but it is not necessary for constructing phyloseq
object. Construct Phyloseq object called "Physeq"

```{r warning=FALSE}
sample_names(OTU)
sample_names(TAX)
sample_names(sampleData)


physeq = phyloseq(OTU,TAX,sampleData)
```

Check phyloseq object This should indicate that your physeq is a
"phyloseq-class experiment-level object""

```{r warning=FALSE}
physeq
```

###STEP9: Remove singletons Remove any OTUs that present only one time.

```{r }
physeq.prune = prune_taxa(taxa_sums(physeq) > 1, physeq)
```

```{r warning=FALSE}
physeq.prune
```

###STEP10: Plot read counts to check dataset 
Check read counts: any samples that have very low reads should be removed.
[Ref](http://evomics.org/wp-content/uploads/2016/01/phyloseq-Lab-01-Answers.html)

```{r}
readcount = data.table(as(sample_data(physeq.prune), "data.frame"),
                 TotalReads = sample_sums(physeq.prune), 
                 keep.rownames = TRUE)
setnames(readcount, "rn", "SampleID")

```

For plotting, use command below.
```{r}
#SeqDepth = ggplot(readcount, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
#SeqDepth
#ggsave("./Figures_glomeromycetes/SeqDepth_plot.png", plot = SeqDepth, width = 6, height = 4, units = "in")
```


TotalReads of all the samples can be in this table (select only SampleID
and TotalReads columns). In order to check samples with low number of
reads, "order()" can be used to sort "TotalReads" column. In this
dataset, N55.Rhizo has very low number of reads, so will will filter
this sample out using the next minimum number of reads.

```{r}
readcount = readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")]
```

```{r}
head(readcount)
```

The following code rarefys the data. Make sure the sample size is equal to the sample with the lowest read count. 
```{r warning=FALSE}
set.seed(1)
physeq.prune.rarefy = rarefy_even_depth(physeq.prune, sample.size = 7909, replace = FALSE, trimOTUs = FALSE)
physeq.prune.rarefy
```


###STEP11 Taxonomic composition Glomeromyectes (melted)

```{r}
physeq.prune.rarefy.Glomeromycetes = subset_taxa(physeq.prune.rarefy, Class == "Glomeromycetes")
physeq.prune.rarefy.Glomeromycetes
```

```{r}
# Get the Dali palette
Dali <- moma.colors("Dali")

# Add one more color to the palette
new_color <- "black"  # Example color, you can change it to any color code you like
Dali <- c(Dali, new_color)
```


```{r fig.height=5, fig.width=8, fig.align="center"}
library(dplyr)

taxcom_Substrate_glom <- ggplot(data = psmelt(physeq.prune.rarefy.Glomeromycetes) %>%
                                    filter(!is.na(Genus)),  # Filter out rows where Genus is NA
                                  mapping = aes_string(x = "Substrate", y = "Abundance", fill = "Genus")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  
  ggtitle("Glomeromycete Taxonomy (Genus) by Substrate") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = Dali) +  
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_Substrate_glom)

```




```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures_glomeromycetes/Substrate_Glomeromycete_Melted_Taxonomic_Composition_(Genus_level).pdf", width = 8, height = 5.8 )
taxcom_Substrate_glom
dev.off()
```



###STEP14 Taxonomic composition Glomeromyectes (unmelted) by site

####Start with Tukey

```{r}
sample_data(physeq.prune.rarefy.Glomeromycetes)$Site = factor(sample_data(physeq.prune.rarefy.Glomeromycetes)$Site, levels = c("CIMA", "ODLO", "AB", "GMT","TP", "TP2"))

```

```{r}
sums.Glomeromycetes = sample_sums(physeq.prune.rarefy.Glomeromycetes)
sums.Glomeromycetes
z <- data.frame(sums.Glomeromycetes)
z
```

```{r}

data.anova.Glom = cbind(sample_data(physeq.prune.rarefy.Glomeromycetes), sums.Glomeromycetes)
physeq.prune.rarefy.anova.Glomeromycetes = aov(sums.Glomeromycetes
 ~ Site, data.anova.Glom)
summary(physeq.prune.rarefy.anova.Glomeromycetes)
```

```{r}
tukey.ps = TukeyHSD(x=physeq.prune.rarefy.anova.Glomeromycetes, 'Site', conf.level = 0.95)

generate_label_df <- function(tukey.ps, variable){
  
  # Extract labels and factor levels from Tukey post-hoc 
  Tukey.levels <- tukey.ps[[variable]][,4]
  Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])
  
  #I need to put the labels in the same order as in the boxplot :
  Tukey.labels$treatment=rownames(Tukey.labels)
  Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
  return(Tukey.labels)
}

LABELS=generate_label_df(tukey.ps, 'Site')
names(LABELS) = c('Letters','Site')
```

```{r}
LABELS
```

```{r}
ylabel <- data.frame("ylabel" = c(500,500,500,500,3600, 500))
LABELS$ylabel<-ylabel$ylabel
Class <- data.frame("Class" = c("Glomeromycetes"))
LABELS$Class<-Class$Class
LABELS
```



Plot diversity

```{r fig.height=5, fig.width=8, fig.align="center"}
color <- c("lightblue")
tax_Substrate_class_rhodo_abundance <- plot_bar(physeq.prune.rarefy.Glomeromycetes, x="Site", fill = "Class") + 
  geom_bar(aes(color=Class, fill=Class), stat="identity", position="stack", color="transparent") +  # Set color to "transparent" to remove borders
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = color) +  # Set high contrast colors 
  ggtitle("Glomeromycete relative abundance by site") +
  stat_compare_means(method = "anova", label.y = 3000) + geom_text(data=LABELS, aes(x=LABELS$Site, y = LABELS$ylabel , label = LABELS$Letters))

  
# Display the plot
tax_Substrate_class_rhodo_abundance

```
```{r fig.height=5, fig.width=8, fig.align="center"}

pdf("./Figures_glomeromycetes/Glomeromycete_abundance_Site.pdf", width = 8, height = 5 )
tax_Substrate_class_rhodo_abundance
```


###STEP15 Taxonomic composition Glomeromyectes (unmelted) by Substrate

```{r fig.height=5, fig.width=8, fig.align="center"}

tax_Substrate_class_rhodo_abundance <- plot_bar(physeq.prune.rarefy.Glomeromycetes, x="Substrate", fill = "Genus")  + 
  geom_bar(aes(color=Species, fill=Species), stat="identity", position="stack", color="transparent") +  # Set color to "transparent" to remove borders
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = Dali) +  # Set high contrast colors 
  ggtitle("Rhodotorula Taxonomy (Species) by Substrate")

# Display the plot
tax_Substrate_class_rhodo_abundance

```

###STEP16 Taxonomic composition Glomeromyectes (unmelted) by Climate


```{r fig.height=5, fig.width=8, fig.align="center"}

tax_Substrate_class_rhodo_abundance <- plot_bar(physeq.prune.rarefy.Glomeromycetes, x="Climate", fill = "Genus") + 
  geom_bar(aes(color=Species, fill=Species), stat="identity", position="stack", color="transparent") +  # Set color to "transparent" to remove borders
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = Dali) +  # Set high contrast colors 
  ggtitle("Rhodotorula Taxonomy (Species) by Substrate") 

# Display the plot
tax_Substrate_class_rhodo_abundance

```

###STEP17 Taxonomic composition Rhizophagus (unmelted) by Substrate

```{r}

physeq.prune.rarefy.Rhizo = subset_taxa(physeq.prune.rarefy, Genus == "Rhizophagus")
physeq.prune.rarefy.Rhizo
```
####Start with Tukey

```{r}
sample_data(physeq.prune.rarefy.Rhizo)$Substrate = factor(sample_data(physeq.prune.rarefy.Rhizo)$Substrate, levels = c("SMC.ster", "SMC", "Dirt", "SMC.sub"))

```

```{r}
sums.Glomeromycetes = sample_sums(physeq.prune.rarefy.Rhizo)
sums.Glomeromycetes
z <- data.frame(sums.Glomeromycetes)
z
```

```{r}

data.anova.Glom = cbind(sample_data(physeq.prune.rarefy.Rhizo), sums.Glomeromycetes)
physeq.prune.rarefy.anova.Glomeromycetes = aov(sums.Glomeromycetes
 ~ Substrate, data.anova.Glom)
summary(physeq.prune.rarefy.anova.Glomeromycetes)
```

```{r}
tukey.ps = TukeyHSD(x=physeq.prune.rarefy.anova.Glomeromycetes, 'Substrate', conf.level = 0.95)

generate_label_df <- function(tukey.ps, variable){
  
  # Extract labels and factor levels from Tukey post-hoc 
  Tukey.levels <- tukey.ps[[variable]][,4]
  Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])
  
  #I need to put the labels in the same order as in the boxplot :
  Tukey.labels$treatment=rownames(Tukey.labels)
  Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
  return(Tukey.labels)
}

LABELS=generate_label_df(tukey.ps, 'Substrate')
names(LABELS) = c('Letters','Substrate')
```

```{r}
LABELS
```

```{r}
ylabel <- data.frame("ylabel" = c(50,10,50,5))
LABELS$ylabel<-ylabel$ylabel
Class <- data.frame("Class" = c("Glomeromycetes"))
LABELS$Class<-Class$Class
LABELS
```

####Plot

```{r fig.height=5, fig.width=8, fig.align="center"}

tax_Substrate_class_rhizo_abundance <- plot_bar(physeq.prune.rarefy.Rhizo, x="Substrate", fill = "Class")  + 
  geom_bar(aes(color=Genus, fill=Genus), stat="identity", position="stack", color="transparent") +  # Set color to "transparent" to remove borders
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = c("black","#6ea8ab")) +  # Set high contrast colors 
  ggtitle("Rhizophagus by Substrate")  +
  stat_compare_means(method = "anova", label.y = 60) + geom_text(data=LABELS, aes(x=LABELS$Substrate, y = LABELS$ylabel , label = LABELS$Letters))

# Display the plot
tax_Substrate_class_rhizo_abundance

```

```{r fig.height=5, fig.width=8, fig.align="center"}

pdf("./Figures_glomeromycetes/Rhizophagus_abundance_Substrate.pdf", width = 8, height = 5 )
tax_Substrate_class_rhizo_abundance
```

###STEP18 Taxonomic composition Glomus (unmelted) by Substrate

```{r}

physeq.prune.rarefy.Glomus = subset_taxa(physeq.prune.rarefy, Genus == "Glomus")
physeq.prune.rarefy.Glomus
```
####Start with Tukey

```{r}
sample_data(physeq.prune.rarefy.Glomus)$Site = factor(sample_data(physeq.prune.rarefy.Glomus)$Substrate, levels = c("SMC.ster", "SMC", "Dirt", "SMC.sub"))

```

```{r}
sums.Glomeromycetes = sample_sums(physeq.prune.rarefy.Glomus)
sums.Glomeromycetes
z <- data.frame(sums.Glomeromycetes)
z
```

```{r}

data.anova.Glom = cbind(sample_data(physeq.prune.rarefy.Glomus), sums.Glomeromycetes)
physeq.prune.rarefy.anova.Glomeromycetes = aov(sums.Glomeromycetes
 ~ Substrate, data.anova.Glom)
summary(physeq.prune.rarefy.anova.Glomeromycetes)
```

```{r}
tukey.ps = TukeyHSD(x=physeq.prune.rarefy.anova.Glomeromycetes, 'Substrate', conf.level = 0.95)

generate_label_df <- function(tukey.ps, variable){
  
  # Extract labels and factor levels from Tukey post-hoc 
  Tukey.levels <- tukey.ps[[variable]][,4]
  Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])
  
  #I need to put the labels in the same order as in the boxplot :
  Tukey.labels$treatment=rownames(Tukey.labels)
  Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
  return(Tukey.labels)
}

LABELS=generate_label_df(tukey.ps, 'Substrate')
names(LABELS) = c('Letters','Substrate')
```

```{r}
LABELS
```

```{r}
ylabel <- data.frame("ylabel" = c(20,20,230,50))
LABELS$ylabel<-ylabel$ylabel
Class <- data.frame("Class" = c("Glomeromycetes"))
LABELS$Class<-Class$Class
LABELS
```


```{r fig.height=5, fig.width=8, fig.align="center"}

tax_Substrate_class_glomus_abundance <- plot_bar(physeq.prune.rarefy.Glomus, x="Substrate", fill = "Class")  + 
  geom_bar(aes(color=Species, fill=Species), stat="identity", position="stack", color="transparent") +  # Set color to "transparent" to remove borders
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = c("black","#6ea8ab")) +  # Set high contrast colors 
  ggtitle("Glomus by Substrate")  +
  stat_compare_means(method = "anova", label.y = 250) + geom_text(data=LABELS, aes(x=LABELS$Substrate, y = LABELS$ylabel , label = LABELS$Letters))

# Display the plot
tax_Substrate_class_glomus_abundance

```

```{r fig.height=5, fig.width=8, fig.align="center"}

pdf("./Figures_glomeromycetes/glomus_abundance_Substrate.pdf", width = 8, height = 5 )
tax_Substrate_class_glomus_abundance
```
