---
title: "Mojave_Colorado_Div_ITS_phyloseq"
author: "Mostly Nuttapon Pombupba, a bit by Kian Kelly"
date: "Updated on December 21, 2023"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###STEP1: Load all necessary packages

```{r warning=FALSE, message=FALSE}
library(ape)
library(vegan)
library(dplyr)
library(scales)
library(grid)
library(reshape2)
library(phyloseq)
library(magrittr)
library(ggplot2)
library(ggpubr)
library(plyr)
library(data.table)
library(tidyr)
library(tidyverse)
library(multcompView)
```

####Initiallize functions
This function was written by Joey (but it is not part of phyloseq
package) <https://github.com/joey711/phyloseq/issues/418>

Kian notes: This defines functions which will be used in future steps. fast_melt takes an input which is a physeq object, extracts OTU table, modifies this (ie removes NA values). summarize_taxa summarizes the relative abundance data by taxonomic rank, but can also be used with grouping variables (ie crust type) and can calculate summary stats. The result is a data.table containing mean, standard deviation, minimum, and maximum relative abundances for each taxonomic group.

```{r warning=FALSE}
fast_melt = function(physeq){
  # supports "naked" otu_table as `physeq` input.
  otutab = as(otu_table(physeq), "matrix")
  if(!taxa_are_rows(physeq)){otutab <- t(otutab)}
  otudt = data.table(otutab, keep.rownames = TRUE)
  setnames(otudt, "rn", "taxaID")
  # Enforce character taxaID key
  otudt[, taxaIDchar := as.character(taxaID)]
  otudt[, taxaID := NULL]
  setnames(otudt, "taxaIDchar", "taxaID")
  # Melt count table
  mdt = melt.data.table(otudt, 
                        id.vars = "taxaID",
                        variable.name = "SampleID",
                        value.name = "count")
  # Remove zeroes, NAs
  mdt <- mdt[count > 0][!is.na(count)]
  # Calculate relative abundance
  mdt[, RelativeAbundance := count / sum(count), by = SampleID]
  if(!is.null(tax_table(physeq, errorIfNULL = FALSE))){
    # If there is a tax_table, join with it. Otherwise, skip this join.
    taxdt = data.table(as(tax_table(physeq, errorIfNULL = TRUE), "matrix"), keep.rownames = TRUE)
    setnames(taxdt, "rn", "taxaID")
    # Enforce character taxaID key
    taxdt[, taxaIDchar := as.character(taxaID)]
    taxdt[, taxaID := NULL]
    setnames(taxdt, "taxaIDchar", "taxaID")
    # Join with tax table
    setkey(taxdt, "taxaID")
    setkey(mdt, "taxaID")
    mdt <- taxdt[mdt]
  }
  return(mdt)
}

summarize_taxa = function(physeq, Rank, GroupBy = NULL){
  Rank <- Rank[1]
  if(!Rank %in% rank_names(physeq)){
    message("The argument to `Rank` was:\n", Rank,
            "\nBut it was not found among taxonomic ranks:\n",
            paste0(rank_names(physeq), collapse = ", "), "\n",
            "Please check the list shown above and try again.")
  }
  if(!is.null(GroupBy)){
    GroupBy <- GroupBy[1]
    if(!GroupBy %in% sample_variables(physeq)){
      message("The argument to `GroupBy` was:\n", GroupBy,
              "\nBut it was not found among sample variables:\n",
              paste0(sample_variables(physeq), collapse = ", "), "\n",
              "Please check the list shown above and try again.")
    }
  }
  # Start with fast melt
  mdt = fast_melt(physeq)
  if(!is.null(GroupBy)){
    # Add the variable indicated in `GroupBy`, if provided.
    sdt = data.table(SampleID = sample_names(physeq),
                     var1 = get_variable(physeq, GroupBy))
    setnames(sdt, "var1", GroupBy)
    # Join
    setkey(sdt, SampleID)
    setkey(mdt, SampleID)
    mdt <- sdt[mdt]
  }
  # Summarize
  Nsamples = nsamples(physeq)
  summarydt = mdt[, list(meanRA = sum(RelativeAbundance)/Nsamples,
                         sdRA = sd(RelativeAbundance),
                         minRA = min(RelativeAbundance),
                         maxRA = max(RelativeAbundance)),
                  by = c(Rank, GroupBy)]
  return(summarydt)
}

plot_taxa_summary = function(physeq, Rank, GroupBy = NULL){
  # Get taxa summary table 
  dt1 = summarize_taxa(physeq, Rank = Rank, GroupBy = GroupBy)
  # Set factor appropriately for plotting
  RankCol = which(colnames(dt1) == Rank)
  setorder(dt1, -meanRA)
  dt1[, RankFac := factor(dt1[[Rank]], 
                          levels = rev(dt1[[Rank]]))]
  dt1[, ebarMax := max(c(0, min(meanRA + sdRA))), by = eval(Rank)]
  dt1[, ebarMin := max(c(0, min(meanRA - sdRA))), by = eval(Rank)]
  # Set zeroes to one-tenth the smallest value
  ebarMinFloor = dt1[(ebarMin > 0), min(ebarMin)]
  ebarMinFloor <- ebarMinFloor / 10
  dt1[(ebarMin == 0), ebarMin := ebarMinFloor]
  
  pRank = ggplot(dt1, aes(x = meanRA, y = RankFac)) +
    scale_x_log10() +
    xlab("Mean Relative Abundance") +
    ylab(Rank) +
    theme_bw()
  if(!is.null(GroupBy)){
    # pRank <- pRank + facet_wrap(facets = as.formula(paste("~", GroupBy)))
    pRank <- pRank + geom_point(mapping = aes_string(colour = GroupBy),
                                size = 5)
  } else {
    # Don't include error bars for faceted version
    pRank <- pRank + geom_errorbarh(aes(xmax = ebarMax,
                                        xmin = ebarMin))
  }
  return(pRank)
}
```

###STEP2: Import Mapping file (metadate file) 

Kian notes:

I excluded Nat's samples from this analysis by removing them from the metadata file. 

```{r}
meta = read.table("metadata_khk.tsv",header=TRUE,row.names=1,sep="\t",stringsAsFactors=FALSE)
```

###STEP3: Check if your metadata file has been imported successfully 

```{r warning=FALSE}
head(meta)
```

###STEP4: Construct sample_data-class using imported metadata

Kian notes: basically makes a phyloseq version of your metadata file
```{r}
sampleData <- sample_data(meta)
```

###STEP5: Import OTU table
Kian notes: OTU table here combines my data with Nat's.
```{r}
getwd()
otus <- read.table("Mojave2020_KKelly202307.otu_table.txt",header=T,sep="\t",row.names=1)
otumat <- as(as.matrix(otus), "matrix")
OTU = otu_table(otumat, taxa_are_rows = TRUE)
```

Check imported OTU table

```{r warning=FALSE}
head(OTU)
```

###STEP6: Import taxonomy table Taxonmy 
Note: table generated from AMPtk needs to be rearranged using a perl script.

"perl rdp_taxonmy2mat.pl\<Input_taxonmy.txt\>Output_taxonomy.txt"

rdp_taxonomy2mat.pl was created by Professor Jason E. Stajich

```{r}
taxmat <- read.table("Mojave2020_KKelly202307.ASVs.taxonomy.fix.txt", header=T,sep="\t",row.names=1)
taxmat <- as(as.matrix(taxmat),"matrix")
TAX = tax_table(taxmat)
head(TAX)
```

###STEP7: Import phylogenetic tree

```{r warning=FALSE}
treefile = "Mojave2020_KKelly202307.ASVs.tree.phy"
tree = read.tree(treefile)
```

###STEP8: Construct Phyloseq object 

To construct phyloseq object, otu
table, taxonomy table, and sampleData are required. Phylogenetic tree
can be included, but it is not necessary for constructing phyloseq
object. Construct Phyloseq object called "Physeq"

```{r warning=FALSE}
physeq = phyloseq(OTU,TAX,sampleData)
```

Check phyloseq object This should indicate that your physeq is a
"phyloseq-class experiment-level object""

```{r warning=FALSE}
physeq
```

###STEP9: Remove singletons Remove any OTUs that present only one time.

```{r }
physeq.prune = prune_taxa(taxa_sums(physeq) > 1, physeq)
```

```{r warning=FALSE}
physeq.prune
```

###STEP10: Plot read counts to check dataset 
Check read counts: any samples that have very low reads should be removed.
[Ref](http://evomics.org/wp-content/uploads/2016/01/phyloseq-Lab-01-Answers.html)

```{r}
readcount = data.table(as(sample_data(physeq.prune), "data.frame"),
                 TotalReads = sample_sums(physeq.prune), 
                 keep.rownames = TRUE)
setnames(readcount, "rn", "SampleID")

```

For plotting, use command below.
```{r}
SeqDepth = ggplot(readcount, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth")
SeqDepth
ggsave("./Figures/SeqDepth_plot.png", plot = SeqDepth, width = 6, height = 4, units = "in")
```


TotalReads of all the samples can be in this table (select only SampleID
and TotalReads columns). In order to check samples with low number of
reads, "order()" can be used to sort "TotalReads" column. In this
dataset, N55.Rhizo has very low number of reads, so will will filter
this sample out using the next minimum number of reads.

```{r}
readcount = readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")]
```

```{r}
head(readcount)
```

The following code rarefys the data. Make sure the sample size is equal to the sample with the lowest read count. 
```{r warning=FALSE}
set.seed(1)
physeq.prune.rarefy = rarefy_even_depth(physeq.prune, sample.size = 7909, replace = FALSE, trimOTUs = FALSE)
physeq.prune.rarefy
```

###STEP11: Rarefaction curve 
Extract OTU table, transpose and convert to Extract OTU table from phyloseq object as data frame and use vegan to calculate rarefaction curve

```{r}
otu.rare <- otu_table(physeq.prune.rarefy)
otu.rare <- as.data.frame(t(otu.rare))
sample_names <- rownames(otu.rare)

#vegan rarecurve 
otu.rarecurve <- rarecurve(otu.rare, step = 500, sample = 30000, label = T)
```
this code takes a list of rarefaction curves (otu.rarecurve), converts each curve into a data frame, creates a new data frame with columns "OTU" and "Sample.size," adjusts the format of the "Sample.size" column, and returns a list of modified data frames (rare). The goal is to organize and preprocess the data for furthervisualization.

```{r}
rare <- lapply(otu.rarecurve, function(x){
  b <- as.data.frame(x)
  b <- data.frame(OTU = b[,1], Sample.size = rownames(b))
  b$Sample.size <- as.numeric(gsub("N", "",  b$Sample.size))
  return(b)
})
```

label list

```{r}
names(rare) <- sample_names
```

convert to data frame

```{r}
#tidyverse
rare <- map_dfr(rare, function(x){
  z <- data.frame(x)
  return(z)
}, .id = "sample")
```

```{r}
head(rare)
```

Plot rarecurve using ggplot

```{r}
ggplot(data = rare)+
  geom_line(aes(x = Sample.size, y = OTU, color = sample), show.legend = FALSE)+
  scale_x_continuous(labels =  waiver()) + theme_bw() + ggtitle("Fungal Rarefaction Curve")+
  theme(plot.title = element_text(hjust = 0.5))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
```

Save ggplot as pdf to figures folder using the full path

```{r}
pdf("./Figures/Fungal Rarefaction Curve.pdf", width = 8, height = 5 )
ggplot(data = rare)+
  geom_line(aes(x = Sample.size, y = OTU, color = sample), show.legend = FALSE)+
  scale_x_continuous(labels =  waiver()) + theme_bw() + ggtitle("Fungal Rarefaction Curve")+
  theme(plot.title = element_text(hjust = 0.5))
dev.off()
```

###STEP11.1: Plot Fungal Alpha diversity by Crust_type (Chao1 and Observed)

Alpha diversity can be
Chao1, Observed, Shannon, Simpson This plot include statistical analysis
using "stat_compare_means" with "method = anova"


```{r}
physeq.prune.rarefy.plot.richness.Crust_type = plot_richness(physeq.prune.rarefy, x="Crust_type", color=("Crust_type"), measures=c("Chao1")) + geom_boxplot() + ggtitle("Mojave Biocrust Fungal Alpha Diversity (Chao1) by Crust type") + stat_compare_means(method = "anova", label.y = 3000) + stat_compare_means(aes(label=..p.signif..), method = "t.test", ref.group = "LAC", label.y = 2800, hide.ns = TRUE) + theme_bw() + theme(plot.title = element_text(hjust = 0.5)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

physeq.prune.rarefy.plot.richness.Crust_type
```
```{r}
pdf("./Figures/Mojave Biocrust Fungal Alpha Diversity (Chao1) by Crust type.pdf", width = 8, height = 5)
physeq.prune.rarefy.plot.richness.Crust_type
dev.off()
```

```{r}
physeq.prune.rarefy.plot.richness.Crust_type = plot_richness(physeq.prune.rarefy, x="Crust_type", color=("Crust_type"), measures=c("Observed")) + geom_boxplot() + ggtitle("Mojave Biocrust Fungal Alpha Diversity (Observed) by Crust type") + stat_compare_means(method = "anova", label.y = 3000) + stat_compare_means(aes(label=..p.signif..), method = "t.test", ref.group = "LAC", label.y = 2800, hide.ns = TRUE) + theme_bw() + theme(plot.title = element_text(hjust = 0.5)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

physeq.prune.rarefy.plot.richness.Crust_type


```

```{r}
pdf("./Figures/Mojave Biocrust Fungal Alpha Diversity (Observed) by Crust type.pdf", width = 8, height = 5)
physeq.prune.rarefy.plot.richness.Crust_type
dev.off()
```

###STEP11.3: Plot Fungal Alpha diversity by Site (Chao1 and Observed)

First perform Tukey test, then plot aplha div

```{r}
physeq.prune.rarefy.SF = subset_samples(physeq.prune.rarefy, Layer=="Surface")
```


```{r}
physeq.prune.rarefy.SF = subset_samples(physeq.prune.rarefy, Layer=="Surface")
sample_data(physeq.prune.rarefy.SF)$Site = factor(sample_data(physeq.prune.rarefy.SF)$Site, levels = c("CIMA", "ODLO", "AB"))
```

```{r}
alpha.diversity.SF = estimate_richness(physeq.prune.rarefy.SF, measures = c("Observed"))
data.anova.SF = cbind(sample_data(physeq.prune.rarefy.SF), alpha.diversity.SF)
physeq.prune.rarefy.anova.SF = aov(Observed ~ Site, data.anova.SF)
summary(physeq.prune.rarefy.anova.SF)
```


```{r}
tukey.ps = TukeyHSD(x=physeq.prune.rarefy.anova.SF, 'Site', conf.level = 0.95)

generate_label_df <- function(tukey.ps, variable){
  
  # Extract labels and factor levels from Tukey post-hoc 
  Tukey.levels <- tukey.ps[[variable]][,4]
  Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])
  
  #I need to put the labels in the same order as in the boxplot :
  Tukey.labels$treatment=rownames(Tukey.labels)
  Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
  return(Tukey.labels)
}

LABELS=generate_label_df(tukey.ps, 'Site')
names(LABELS) = c('Letters','Site')
```

```{r}
LABELS
```

```{r}
ylabel <- data.frame("ylabel" = c(245,235,215))
LABELS$ylabel<-ylabel$ylabel
LABELS
```

```{r}
physeq.prune.rarefy.plot.richness.Site = plot_richness(physeq.prune.rarefy.SF, x="Site", color=("Site"), measures=c("Chao1")) + geom_boxplot(lwd=0.5) + ggtitle("Mojave Biocrust Fungal Alpha Diversity by Site") + stat_compare_means(method = "anova", label.y = 280) + theme(plot.title = element_text(hjust = 0.5)) + geom_text(data=LABELS, aes(x=LABELS$Site, y = LABELS$ylabel , label = LABELS$Letters)) + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_colour_brewer(palette="Set1")

physeq.prune.rarefy.plot.richness.Site
```

```{r warning=FALSE}
pdf("./Figures/Fungal_Alpha_Div_by_Site_(Chao1).pdf", width = 6, height = 3.8 )
physeq.prune.rarefy.plot.richness.Site
dev.off()
```

```{r}
physeq.prune.rarefy.plot.richness.Site = plot_richness(physeq.prune.rarefy.SF, x="Site", color=("Site"), measures=c("Observed")) + geom_boxplot(lwd=0.5) + ggtitle("Mojave Biocrust Fungal Alpha Diversity by Site") + stat_compare_means(method = "anova", label.y = 280) + theme(plot.title = element_text(hjust = 0.5)) + geom_text(data=LABELS, aes(x=LABELS$Site, y = LABELS$ylabel , label = LABELS$Letters)) + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_colour_brewer(palette="Set1")
physeq.prune.rarefy.plot.richness.Site
```




```{r warning=FALSE}
pdf("./Figures/Fungal_Alpha_Div_by_Site_(Observed).pdf", width = 6, height = 3.8 )
physeq.prune.rarefy.plot.richness.Site
dev.off()
```


###STEP11.4: Plotting alpha diversity for each class by site
Subset fungal OTUs by class

```{r}
FungiClass = as.character(get_taxa_unique(physeq.prune.rarefy.SF, "Class"))
FungiClass = FungiClass[complete.cases(FungiClass)]
```

Save pdf of plot_richness results for each fungal class in the loop
which go through FungiClass Note: following code plot alpha diversity by
Site and Layer

```{r}
pdf("./Figures/Fungal_Class_Alpha_Div_by_Site.pdf", width = 8, height = 5)

for (i in FungiClass) {
  APS = subset_taxa(physeq.prune.rarefy.SF, Class == i)
  APS.plot = plot_richness(APS, x="Site", color =("Site"), measures=c("Observed", "Chao1")) +
    ggtitle(i) +
    stat_compare_means(method = "anova") +
    theme_bw() +
    geom_boxplot() +
    theme(plot.title = element_text(hjust = 0.5))
  print(APS.plot)
}

dev.off()

```
###STEP11.5: Plotting alpha diversity for each class by crust type


Save pdf of plot_richness results for each fungal class in the loop
which go through FungiClass Note: following code plot alpha diversity by
Crust type

```{r}
pdf("./Figures/Fungal_Class_Alpha_Div_by_Crust_type.pdf", width = 8, height = 5)

for (i in FungiClass) {
  APS = subset_taxa(physeq.prune.rarefy.SF, Class == i)
  APS.plot = plot_richness(APS, x="Crust_type", color =("Crust_type"), measures=c("Observed", "Chao1")) +
    ggtitle(i) +
    stat_compare_means(method = "anova") +
    theme_bw() +
    geom_boxplot() +
    theme(plot.title = element_text(hjust = 0.5))
  print(APS.plot)
}

dev.off()

```

###STEP11.6: Plotting alpha diversity for each genus by crust type

```{r}
FungiGenus = as.character(get_taxa_unique(physeq.prune.rarefy.SF, "Genus"))
FungiGenus = FungiGenus[complete.cases(FungiGenus)]
```

```{r}
pdf("./Figures/Fungal_Genus_Alpha_Div_by_Crust_type.pdf", width = 8, height = 5)

for (i in FungiGenus) {
  APS = subset_taxa(physeq.prune.rarefy.SF, Genus == i)
  APS.plot = plot_richness(APS, x="Crust_type", color =("Crust_type"), measures=c("Observed", "Chao1")) +
    ggtitle(i) +
    stat_compare_means(method = "anova") +
    theme_bw() +
    geom_boxplot() +
    theme(plot.title = element_text(hjust = 0.5))
  print(APS.plot)
}

dev.off()

```
###STEP11.7: Plotting alpha diversity for each genus by site and layer



```{r}
pdf("./Figures/Fungal_Genus_Alpha_Div_by_Site.pdf", width = 8, height = 5)

for (i in FungiGenus) {
  APS = subset_taxa(physeq.prune.rarefy.SF, Genus == i)
  APS.plot = plot_richness(APS, x="Site", color =("Site"), measures=c("Observed", "Chao1")) +
    ggtitle(i) +
    stat_compare_means(method = "anova") +
    theme_bw() +
    geom_boxplot() +
    theme(plot.title = element_text(hjust = 0.5))
  print(APS.plot)
}

dev.off()

```
```{r}
pdf("./Figures/Fungal_Genus_Alpha_Div_by_Layer.pdf", width = 8, height = 5)

for (i in FungiGenus) {
  APS = subset_taxa(physeq.prune.rarefy, Genus == i)
  APS.plot = plot_richness(APS, x="Layer", color =("Layer"), measures=c("Observed", "Chao1")) +
    ggtitle(i) +
    stat_compare_means(method = "anova") +
    theme_bw() +
    geom_boxplot() +
    theme(plot.title = element_text(hjust = 0.5))
  print(APS.plot)
}

dev.off()

```


###STEP11.8: Plotting alpha diversity for each class by layer

```{r}
sample_data(physeq.prune.rarefy)$Layer = factor(sample_data(physeq.prune.rarefy)$Layer, levels = c("Surface", "Subsurface"))
```

```{r}
FungiClass.all = as.character(get_taxa_unique(physeq.prune.rarefy, "Class"))
FungiClass.all = FungiClass.all[complete.cases(FungiClass.all)]
FungiClass.all
```

```{r}
pdf("./Figures/Fungal_class_Alpha_Div_by_Layer.pdf", width = 8, height = 5)

for (i in FungiClass.all) {
  APS = subset_taxa(physeq.prune.rarefy, Class == i)
  APS.plot = plot_richness(APS, x="Layer", color =("Layer"), measures=c("Observed", "Chao1")) +
    ggtitle(i) +
    stat_compare_means(method = "anova") +
    theme_bw() +
    geom_boxplot() +
    theme(plot.title = element_text(hjust = 0.5))
  print(APS.plot)
}

dev.off()

```

###STEP12.1 Taxonomic composition (ugly bar plots)


```{r}
taxa.summary.by.phylum = summarize_taxa(physeq.prune.rarefy, "Phylum")
```

######To save taxa summary into text file under your current working directory.

```{r warning=FALSE}
write.table(taxa.summary.by.phylum, file = "taxa.summary.by.phylum.txt", sep ="\t")
```

```{r}
physeq.prune.rarefy
```

```{r}
physeq.prune.rarefy = subset_taxa(physeq.prune.rarefy, Kingdom == "Fungi")
physeq.prune.rarefy
```

By class, layer

```{r}
ps.glom.class = tax_glom(physeq.prune.rarefy, taxrank = "Class")
```

```{r}
sample_data(physeq.prune.rarefy)$Layer = factor(sample_data(physeq.prune.rarefy)$Layer, levels = c("Surface", "Subsurface"))
```


```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_crust_type <- ggplot(data = psmelt(physeq.prune.rarefy), 
                            mapping = aes_string(x = "Crust_type", y = "Abundance", fill = "Class")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Fungal Taxonomy (Class) by Crust_type") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_crust_type)
```

```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Crust_type_Fungal_Taxonomic_Composition_(Class_level)_by_.pdf", width = 8, height = 5.8 )
taxcom_crust_type
dev.off()
```


```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_crust_site <- ggplot(data = psmelt(physeq.prune.rarefy), 
                            mapping = aes_string(x = "Site", y = "Abundance", fill = "Class")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Fungal Taxonomy (Class) by Site") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_crust_site)
```

```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Site_Fungal_Taxonomic_Composition_(Class_level)_by_site_.pdf", width = 8, height = 5.8 )
taxcom_crust_site
dev.off()
```


```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_crust_layer <- ggplot(data = psmelt(physeq.prune.rarefy), 
                            mapping = aes_string(x = "Layer", y = "Abundance", fill = "Class")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Fungal Taxonomy (Class) by Layer") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_crust_layer)
```

```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Crust_layer_(abundance)_Fungal_Taxonomic_Composition_(Class_level).pdf", width = 8, height = 5.8 )
taxcom_crust_layer
dev.off()
```


###STEP12.2 Taxonomic composition (glom)

```{r}
physeq.prune.rarefy.Glomeromycetes = subset_taxa(physeq.prune.rarefy, Class == "Glomeromycetes")
physeq.prune.rarefy.Glomeromycetes
```


```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_crust_type_glom <- ggplot(data = psmelt(physeq.prune.rarefy.Glomeromycetes), 
                            mapping = aes_string(x = "Crust_type", y = "Abundance", fill = "Genus")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Glomeromycete Taxonomy (Genus) by Crust_type") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_crust_type_glom)
```
```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Crust_type_Glomeromycete_Taxonomic_Composition_(Genus_level).pdf", width = 8, height = 5.8 )
taxcom_crust_type_glom
dev.off()
```

```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_site_glom <- ggplot(data = psmelt(physeq.prune.rarefy.Glomeromycetes), 
                            mapping = aes_string(x = "Site", y = "Abundance", fill = "Genus")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Glomeromycete Taxonomy (Genus) by Site") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_site_glom)
```


```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Site_glom_bar.pdf", width = 8, height = 5.8 )
taxcom_site_glom
dev.off()
```

```{r fig.height=5, fig.width=8, fig.align="center"}
tax_crust_type_class_glom_abundance = plot_bar(physeq.prune.rarefy.Glomeromycetes, x="Crust_type", fill = "Genus") + 
  geom_bar(aes(color=Genus, fill=Genus), stat="identity", position="stack") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Glomeromycete Taxonomy (Genus) by Crust_type")
tax_crust_type_class_glom_abundance
```
```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Crust_type_Glomeromycete_Taxonomic_Composition_(Genus_level)_abundance.pdf", width = 8, height = 5.8 )
tax_crust_type_class_glom_abundance
dev.off()
```


```{r fig.height=5, fig.width=8, fig.align="center"}
tax_site_class_glom = plot_bar(physeq.prune.rarefy.Glomeromycetes, x="Site", fill = "Genus") + 
  geom_bar(aes(color=Genus, fill=Genus), stat="identity", position="stack") +
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Glomeromycete Taxonomy (Genus) by site")
tax_site_class_glom
```
```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Glom_abundance_taxonomy_bar_site.pdf", width = 8, height = 5.8 )
tax_site_class_glom
dev.off()
```

```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_site_glom <- ggplot(data = psmelt(physeq.prune.rarefy.Glomeromycetes), 
                            mapping = aes_string(x = "Site", y = "Abundance", fill = "Genus")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Glomeromycete Taxonomy (Genus) by Site") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_site_glom)
```


```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/Site_glom_bar.pdf", width = 8, height = 5.8 )
taxcom_site_glom
dev.off()
```

###STEP12.3 Taxonomic composition (naganishia)

```{r}
physeq.prune.rarefy.Naga = subset_taxa(physeq.prune.rarefy, Genus == "Naganishia")
physeq.prune.rarefy.Naga
```


```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_crust_type_naga <- ggplot(data = psmelt(physeq.prune.rarefy.Naga), 
                            mapping = aes_string(x = "Crust_type", y = "Abundance", fill = "Species")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Naganishia Taxonomy (Species) by Crust_type") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_crust_type_naga)
```
```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/taxcom_crust_type_naga.pdf", width = 8, height = 5.8 )
taxcom_crust_type_naga
dev.off()
```

```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F", "#999999",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_site_naga <- ggplot(data = psmelt(physeq.prune.rarefy.Naga), 
                            mapping = aes_string(x = "Site", y = "Abundance", fill = "Species")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Naganishia Taxonomy (Species) by Site") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_site_naga)
```


```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/taxcom_site_naga.pdf", width = 8, height = 5.8 )
taxcom_site_naga
dev.off()
```

```{r fig.height=5, fig.width=8, fig.align="center"}
tax_crust_type_class_Naga_abundance = plot_bar(physeq.prune.rarefy.Naga, x="Crust_type", fill = "Species") + 
  geom_bar(aes(color=Species, fill=Species), stat="identity", position="stack") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Naganishia Taxonomy (Species) by Crust_type")
tax_crust_type_class_Naga_abundance
```


```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/tax_crust_type_class_Naga_abundance.pdf", width = 8, height = 5.8 )
tax_crust_type_class_Naga_abundance
dev.off()
```


```{r fig.height=5, fig.width=8, fig.align="center"}
tax_site_class_naga = plot_bar(physeq.prune.rarefy.Naga, x="Site", fill = "Species") + 
  geom_bar(aes(color=Species, fill=Species), stat="identity", position="stack") +
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Naganishia Taxonomy (Species) by site")
tax_site_class_naga
```


```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/tax_site_class_naga.pdf", width = 8, height = 5.8 )
tax_site_class_naga
dev.off()
```


###STEP12.4 Taxonomic composition (Rhodotorula)

```{r}
physeq.prune.rarefy.Rhodo = subset_taxa(physeq.prune.rarefy, Genus == "Rhodotorula")
physeq.prune.rarefy.Rhodo
```


```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_crust_type_rhodo <- ggplot(data = psmelt(physeq.prune.rarefy.Rhodo), 
                            mapping = aes_string(x = "Crust_type", y = "Abundance", fill = "Species")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Rhodotorula Taxonomy (Species) by Crust_type") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_crust_type_rhodo)
```
```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/taxcom_crust_type_rhodo.pdf", width = 8, height = 5.8 )
taxcom_crust_type_rhodo
dev.off()
```

```{r fig.height=5, fig.width=8, fig.align="center"}

# Choose high contrast colors for the bars
high_contrast_colors <- c("#E41A1C", "#377EB8", "#4DAF4A", "#FF7F00", "#F781BF", "#A65628", "#984EA3",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#FFD92F", "#E31A1C", "#1F78B4", "#B2DF8A",
                           "#33A02C", "#FB9A99", "#FF7F00", "#FFFF99", "#4DAF4A", "#F781BF", "#FFD92F",
                           "#66C2A5", "#FC8D62", "#8DD3C7", "#FDAE61", "#E41A1C", "#1F78B4", "#B2DF8A", "#33A02C",
                           "#FB9A99", "#FF7F00", "midnightblue")

taxcom_site_rhodo <- ggplot(data = psmelt(physeq.prune.rarefy.Rhodo), 
                            mapping = aes_string(x = "Site", y = "Abundance", fill = "Species")) + 
  geom_bar(stat = "identity", position = "fill", color = "transparent", linewidth = 0.5) +  # White border for better contrast
  ggtitle("Rhodotorula Taxonomy (Species) by Site") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

print(taxcom_site_rhodo)
```


```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/taxcom_site_rhodo.pdf", width = 8, height = 5.8 )
taxcom_site_rhodo
dev.off()
```

```{r fig.height=5, fig.width=8, fig.align="center"}

tax_crust_type_class_rhodo_abundance <- plot_bar(physeq.prune.rarefy.Rhodo, x="Crust_type", fill = "Species") + 
  geom_bar(aes(color=Species, fill=Species), stat="identity", position="stack", color="transparent") +  # Set color to "transparent" to remove borders
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors 
  ggtitle("Rhodotorula Taxonomy (Species) by Crust_type") 

# Display the plot
tax_crust_type_class_rhodo_abundance

```


```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/tax_crust_type_class_rhodo_abundance.pdf", width = 8, height = 5.8 )
tax_crust_type_class_rhodo_abundance
dev.off()
```


```{r fig.height=5, fig.width=8, fig.align="center"}

tax_site_rhodo_abundance <- plot_bar(physeq.prune.rarefy.Rhodo, x="Site", fill = "Species") + 
  geom_bar(aes(color=Species, fill=Species), stat="identity", position="stack", color="transparent") +  # Set color to "transparent" to remove borders
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_fill_manual(values = high_contrast_colors) +  # Set high contrast colors 
  ggtitle("Rhodotorula Taxonomy (Species) by Site") 

# Display the plot
tax_site_rhodo_abundance

```


```{r fig.height=5, fig.width=8, fig.align="center"}
pdf("./Figures/tax_site_rhodo_abundance.pdf", width = 8, height = 5.8 )
tax_site_rhodo_abundance
dev.off()
```
